### PR TITLE
Explore Map

### PR DESCRIPTION
**Pull request by KhadijaMahanga**
## Description
- Add Explore Map

Fixes # (issue)

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/components/ExploreMap/index.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+function ExploreMap({ center, zoom, styles, geometries, geography, ...props }) {
+  const [selectedBoundary, setSelectedBoundary] = useState(null);
+
+  const getSelectedBoundary = (level, geoms) => {
+    const preferredChildren = preferredChildrenObj[level];
+    if (!preferredChildren) return null;
+
+    const availableLevels = preferredChildren.filter(
+      (l) => geoms.children[l] !== undefined
+    );
```
Comment (kilemensi) [inline src/components/ExploreMap/index.js:22]:
```suggestion
    const availableLevels = preferredChildren.filter((l) => geoms.children[l]);
```

File: src/components/ExploreMap/index.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      zoom={zoom}
+      zoomControl={false}
+      zoomPosition="bottomright"
+      scrollWheelZoom={false}
+      touchZoom={false}
+      zoomSnap={0.25}
+      style={styles}
+    >
+      <Pane name="tiles" style={{ zIndex: 200, pointerEvents: "none" }}>
+        <TileLayer url="https://{s}.basemaps.cartocdn.com/rastertiles/voyager_nolabels/{z}/{x}/{y}.png" />
```
Comment (kilemensi) [inline src/components/ExploreMap/index.js:75]: ENV var

File: src/components/ExplorePage/index.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+}) {
+  const classes = useStyles(props);
+  const [geoCode, setGeoCode] = useState(null);
+  const [geometries, setGeometries] = useState(geometriesProp);
+  const [geography, setGeography] = useState(geographyProp);
+  const [shouldFetch, setShouldFetch] = useState(false);
+
+  const { data } = useSWR(
+    shouldFetch
+      ? `https://ng.hurumap.org/api/v1/all_details/profile/1/geography/${geoCode}/?format=json`
```
Comment (kilemensi) [inline src/components/ExplorePage/index.js:37]: ENV var

File: src/components/ExplorePage/index.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      const g = data?.profile.geography;
+      const geom = {
+        boundary: data?.boundary,
+        children: data?.children, // Dictionary keyed by child type
+        parents: data?.parent_layers ?? [], // Array of parent geographies
+        themes: data?.themes ?? [],
+      };
+      setGeometries(geom);
+      setGeography(g);
+    }
```
Comment (kilemensi) [inline src/components/ExplorePage/index.js:53]:
```suggestion
    if (data) {
      const g = data.profile.geography;
      const geom = {
        boundary: data.boundary,
        children: data.children, // Dictionary keyed by child type
        parents: data.parent_layers ?? [], // Array of parent geographies
        themes: data.themes ?? [],
      };
      setGeometries(geom);
      setGeography(g);
    }
```

We've already checked `if (data)`

File: src/pages/explore/[[...slug]].js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        params: {
+          slug: [s],
+        },
+      };
+    }
+  );
+
+  return {
+    paths,
+    fallback: "blocking",
```
Comment (kilemensi) [inline src/pages/explore/[[...slug]].js:33]: Why?

**Review by kilemensi — COMMENTED** [review]
👍🏾 ... looks good.

----

1. I think lets confirm the map colours i.e. no selection (outside KE), no selection (inside KE), hover, selection (single county).
2. Tooltip needs to match designs

**Comment by kilemensi** [comment]
Forgot to mention: Lets also move component into `src/components/HURUmap` folder and include a story for it @KhadijaMahanga

File: src/pages/explore/[[...slug]].js (hunk 5 above)
Comment (KhadijaMahanga) [reply src/pages/explore/[[...slug]].js:33]: @kilemensi copy & paste..  😆  we don't need this

File: src/components/HURUmap/Map/Layers.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+}) => {
+  const map = useMap();
+  const router = useRouter();
+  const groupRef = useRef();
+
+  const featuredCountiesCode =
+    process.env.NEXT_PUBLIC_FEATURED_COUNTIES?.split(",");
+
+  const popUpContent = (level, name) =>
+    `<div class='tooltip'><div class='level'>${level}</div> <div class='name'>${name.toLowerCase()}</div></div>`;
```
Comment (kilemensi) [inline src/components/HURUmap/Map/Layers.js:59]: Can't this be a React component that we [render as String](https://reactjs.org/docs/react-dom-server.html#rendertostaticmarkup) or something? My thinking being this component is similar/exactly the same used on the map heads-up display of location data.
Screenshot from 2021-09-10 13-34-33

File: src/components/HURUmap/Map/Layers.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+  parentsGeometries,
+  setGeoCode,
+  setShouldFetch,
+}) => {
+  const map = useMap();
+  const router = useRouter();
+  const groupRef = useRef();
+
+  const featuredCountiesCode =
+    process.env.NEXT_PUBLIC_FEATURED_COUNTIES?.split(",");
```
Comment (kilemensi) [inline src/components/HURUmap/Map/Layers.js:56]: 🤔

File: src/components/HURUmap/Map/index.js [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    if (selectedBound?.type === "Feature") {
+      selectedBound = {
+        ...selectedBound,
+        properties: {
+          ...selectedBound.properties,
+          selected: true,
+        },
+      };
+    } else {
+      // else its a featurecollection
```
Comment (kilemensi) [inline src/components/HURUmap/Map/index.js:85]:
```suggestion
      // else its a feature collection
```
Or is it `FeatureCollection`?

File: src/pages/explore/[[...slug]].js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    </Page>
+  );
+}
+
+export async function getStaticPaths() {
+  const result = await fetcher(
+    `${process.env.HURUMAP_API_URL}all_details/profile/1/geography/KE/?format=json`
+  );
+  const featuredCountiesCode =
+    process.env.NEXT_PUBLIC_FEATURED_COUNTIES.split(",");
```
Comment (kilemensi) [inline src/pages/explore/[[...slug]].js:22]: 1. Shouldn't `featuredCountiesCode` be based to Layers and used [here](https://reactjs.org/docs/react-dom-server.html#rendertostaticmarkup)
2. Is this something that is fixed-ish or can change often (ENV vs CMS).

File: src/components/ExplorePage/index.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
+import { Hidden } from "@material-ui/core";
+import { makeStyles } from "@material-ui/core/styles";
+import dynamic from "next/dynamic";
+import PropTypes from "prop-types";
+import React, { useEffect, useState } from "react";
+import useSWR from "swr";
+
+import fetcher from "@/pesayetu/utils/fetcher";
```
Comment (kilemensi) [inline src/components/ExplorePage/index.js:8]: I hear you and SWR use the same name but `fetcher` is weird to me...`fetchJson`, `fetchAsJson`, etc., sound better.

**Review by kilemensi — APPROVED** [review]
🚀 🚀 

-----

Stop showing off!!!
S

File: src/components/HURUmap/Map/Layers.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    if (layer) {
+      layer.clearLayers();
+      const featuredGeo = new L.GeoJSON(selectedBoundary, { onEachFeature });
+      layer.addLayer(featuredGeo);
+      map.fitBounds(layer.getBounds(), {
+        animate: true,
+        duration: 0.5, // in seconds
+      });
+    }
+  }, [groupRef, selectedBoundary]);
```
Comment (kilemensi) [inline src/components/HURUmap/Map/Layers.js:111]: 111:6  Warning: React Hook useEffect has missing dependencies: 'map' and 'onEachFeature'. Either include them or remove the dependency array.

File: src/pages/explore/[[...slug]].js (hunk 9 above)
Comment (KhadijaMahanga) [reply src/pages/explore/[[...slug]].js:22]: @kilemensi confused with the above comments

File: src/components/HURUmap/Map/index.js (hunk 8 above)
Comment (KhadijaMahanga) [reply src/components/HURUmap/Map/index.js:85]: yeah, its a one word. FeatureCollection

File: src/pages/explore/[[...slug]].js (hunk 9 above)
Comment (kilemensi) [reply src/pages/explore/[[...slug]].js:22]: 🤯 ... I meant:

1. We're already loading featuredCountiesCode here and repeating the same thing [here](https://github.com/CodeForAfrica/PesaYetu/pull/89/files#diff-3f05ca11f5b46d1d0d495c906f5b0d0cdd8d267db5d4778e2f72afed9fa7d7b5R55-R56). If they're the same, whey not just pass them as props.
2. Is changing counties something that should involve TECH (then yes, ENV) or is it something that the content/data team will be handling on their own (then CMS).

**Review by KobbyMmo — APPROVED** [review]
looks good to me @KhadijaMahanga

File: src/pages/explore/[[...slug]].js (hunk 9 above)
Comment (KhadijaMahanga) [reply src/pages/explore/[[...slug]].js:21]: @kilemensi ahaa! Simple english now.. yes, let me add them on CMS

**Pull request merged**